luanti-org/luanti
# Support gettext-style language priority list for Minetest's built-in translation system

Add compact, short information about your PR for easier understanding:

(GNU?) Gettext allows users to specify a [priority list of languages](https://www.gnu.org/software/gettext/manual/html_node/The-LANGUAGE-variable.html) (using the `LANGUAGE` variable) that is applied for translations. This can be seen in the screenshot below (using `env LANGUAGE=zh_CN:de minetest`).
![2024-03-03-01-26-00](https://github.com/minetest/minetest/assets/37980625/cdc4be73-cd74-41d5-b72d-68c0ad18ecf5)

The goal of this PR is to make Minetest's own translation system also support this feature.

Note: This PR changes the language code for written Cantonese from `yue_Hant` to `yue`, which breaks systems that use `yue_HK` as the locale. However, considering the (from what I have seen) low translation coverage for this language, I do not think this change will be significant when rolled out.

- Goal of the PR
  See title.
- How does the PR work?
  - The `Translations` class is extended to also hold language information in the lookup table.
  - The translation lookup tables are shared on the server-side to avoid allocating separate translation tables for (combinations of) different languages.
  - `Translations` parses the list of languages and looks up the translation for the languages in the list, returning the first one it finds.
  - The language setting uses GNU gettext's format of having a list of languages separated by a colon. This is parsed by the client as its language setting and by the server to send translations of multiple languages based on the preferences of the client.
  - Backward compatibility is kept:
    - When connecting to an older server from a newer client, the client only sends a fallback language. This allows the existing server-side language detection code to continue to work, but the client only receives translation files for the fallback language.
    - When connecting to a newer server from an older client, the language string from the client is parsed as a list containing a single language.
- Does it resolve any reported issue?
  I can not find a relevant issue.
- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
  Partially 2.2 Internal code refactoring.
  (The description of 2.3 UI Improvements does not seem to suggest that this PR relates to the stated goal.)
- If not a bug fix, why is this PR needed? What usecases does it solve?
  It allows players to specify a list of languages that can be used for translations. In particular, it allows users to specify `zh_CN:zh_TW` or `zh_TW:zh_CN` to use the other language variant.

## To do
This PR is a Work in Progress.

- [x] Extend `Translations` class to also hold language information.
- [x] Implement priority list feature in the `Translations`
- Implement support for the priority list:
  - [x] On the server side (particularly for `minetest.get_translated_string`)
  - [x] On the client side
- Server:
  - [x] Send translation files for multiple languages if this is needed by the client.
- Client:
  - [x] Send priority list of languages to the server.
  - [ ] Allow specifying a fallback language (default: `gettext("LANG_CODE")`?) to send to older servers
  - [x] Send priority list of languages to ContentDB (i.e. follow up #14410)
  - [x] Handle different languages correctly when loading translation files.
  - [x] Extend settings GUI to specify a priority list of languages (see implementation notes).
  - Potential follow-up PR: integrate with Android's "App language" settings.
- Add testcases:
  - [x] `core.get_translated_string`
  - [x] Language priority list expansion.

## How to test
- Server-side translations: covered by `unittests`
- Client-side translations:
  1. Use at least two languages where the primary language has a lower translation coverage than the secondary ones.
  2. Open up a (mod) formspec in a game.
  3. Note that the UI is translated in multiple languages.
An example with the mail mod can be seen in the following screenshot:
![2024-03-05-11-26-47](https://github.com/minetest/minetest/assets/37980625/e81f3120-2fe4-4a69-9bc7-dd531bda5e86)
- Compatibility with older servers: similar to above; note that only the fallback language is used.
- ContentDB: similar to testing with client-side translations, although this is quite subtle to notice as translation coverage for package meta is currently very low.
- Settings menu: manual testing. Note that the list should appear slightly different depending on whether the system language is used and the number of languages in the list.
- Android: in addition to the above:
  1. Add another language to the preferred language order.
  2. Start Minetest with the default language settings.
  4. Uncheck "Use system language".
  5. Note that the languages from the system settings are carried over to MT's language settings if the languages are supported by Minetest.

## Implementation notes
- The language list setting is implemented using its own setting type, `language_list`. While this does make sense in that the language selector has certain cases that need to be handled in a special way:
  - The values are colon-separated entries from a specific list.
  - I have designed the formspec code to also show "invalid" entries.
  - The "empty list" should IMO be displayed differently to indicate that the system language is used. Currently, in addition to the list, there is a "User system language" checkmark that can be used to disable language selection.
  - However, this setting type is also very specific and unlikely to be used elsewhere. Thoughts?
